Request Timeout

ErsinSezgin opened the pull request:
Added missing request timeout to the `worker` and `Client` so that anyone who wants to add timeouts can do it. Moreover, I'd like to add default timeout but I'm not sure about what the value should be.

TODO:

- [x] Decide on timeout value
- [ ] Release the new version of the worker
- [ ] Update current workers incrementally, with the new version

ErsinSezgin committed:
added timeout variable

ErsinSezgin committed:
added timeout to QueueServiceWorker

lindabjalla commented:
I have a question! I would like to know why we want to add timeouts. What happens if we don't add timeouts?

ErsinSezgin commented:
Thanks for the question! We want workers to try to get a job and work on it can complete the task in X amount of minutes. In case something goes wrong, it should crash and die so new freshly built one can take over. 

So every request we make right now assumes server to respond at a given time. This caused `attribution-worker` to stuck at a stage of requested and waited for a response for basically forever. The main problem was Google DNS servers and they were being unable to respond. If we had implemented let's say 10 min timeout, our worker would've been saying `I couldn't do this job put it back to queue and let me try it again` every 10 min and whenever DNS servers are up and running it would've taking over automatically.

We've solved this issue by Eskil going manually killing the worker and causing a creation of a new one. This PR will make sure every worker to crash if they cannot finish the job within given timeout.

For example I'm planning to add something like `timeout=(60, 300)` for our attribution worker. So it will try to do a `handshake` with the server in a min and will wait for a response at tops 5 min. Currently this whole job is done in 2-3 seconds tops so limits are too high, they're for extreme cases.

Hope I addressed your question, let me know if you have another solution in mind or any other questions with current solution :)

alvinlindstam commented:
This client is just used for the workers fetching and reporting for queue items, right?

I like this, but I'm thinking that it might be more suitable to have default or hard coded values for this. I wouldn't mind `timeout=(5, 20)` for all of these. That is plenty long in my mind.

alvinlindstam commented:
Make sure to also do a new version and release for this, using https://github.com/epidemicsound/aws-sqs-worker#release

ErsinSezgin commented:
I totally agree on default value, I'll try to bring it up in Tech Council today to conclude what it should be.

alvinlindstam commented:
I don't think you need to sync it there, nor is it the place to decide such things. Maybe post a suggestion to slack and see if people have opinions?

ErsinSezgin commented:
I was thinking people can proxy the question to their teams, I 100% agree on we should not conclude the decision there but can communicate everyone using it as a proxy. Anyways, I'm writing to slack as well to make it more public. 👍

ErsinSezgin committed:
updated readme

jeppeb91 commented:
Pub/Sub has a max ack deadline of 600 sec! Dunno if this is intended to ever move to pub/sub, tho

edalee reviewed (approved):
You should check with @FreddeR. 
I'm using Pub/Sub now, I don't have any workers here. 
He may have opinions.

pillimoj commented on README.md:
```diff
@@ -77,6 +77,34 @@ an available message in the connected queue. The liveness probe also triggers ev
+queue_name = 'queue_to_pull_messages_from'
+    
+def my_liveness_callback():
+    Path('/worker/alive.txt').touch()
```
Tip. Do not use this with liveness probe..
It runs after every **successful** completion of a task. But the worker might be alive and well even though it has been a while since it got a task that it could process successfully. Also keep in mind that it will only be called after actually finishing so think about how long a task can process at the longest.

If a worker starts receiving payloads with bad data or if another service goes down for example which makes the worker unable to complete the work the worker will be killed. That is probably not what you intended since there is nothing wrong with the worker or it's liveness (this might even happen while the worker is trying to process work that it can process). A liveness check should start failing only once the worker is dead and no longer processing messages.

In the claims-worker we mitigated this by calling "im_alive()" at every start of processing as well.

pillimoj replied:
Maybe add 
```.py
            if self.liveness_callback is not None:
                self.liveness_callback()
```
directly after `while self.run:` in worker.py

ErsinSezgin replied:
Interesting, first of all this is current release 0.0.2 piece that is released by @FreddeR and I actually believe if a worker cannot finish a task within given time or failing it is good to assume that a fresh start is needed. This is my 5 cents of the liveness_probe stuff...

This PR is however is about adding a default timeout and giving an option to configure it depending on your workers need. I think it give only benefits rather than a complexity implementing such thing but I'm not 100% sure what is the best practice considering some workers are actually working on many tasks at the same time

alvinlindstam replied:
@pillimoj, can you tell us some more about 

> the worker might be alive and well even though it has been a while since it got a task that it could process successfully

This would run after each processing of a message (the handler finished without raising an exception) and where the worker successfully cleared the event from the queue, or periodically if there is no messages on the queue. Was your case that the worker continuously needed super long times for the handler or that it continuously raised while you wanted it to be considered healthy?

Trying to understand if we should encourage people to not raise in the handler if they still consider themselves healthy, or if we should move the liveness call to before processing the message in this helper.

pillimoj replied:
In our case an external service was down and the only way to tell the client that the task failed is to throw. The other option would be for the handler to create a duplicate message on the queue if it fails, which feels wrong since the worker is supposed to do the queue stuff and the handler should just execute on the payload if possible.

I would argue for moving the liveness call.
if the handler needs to restart (i.e. its state is bad) it's probably a stateful thing and  should be a service instead.

ErsinSezgin committed:
added default timeout to be (6, 30)
